File: scripts/bump_version.py
#!/usr/bin/env python3
"""Synchronize plugin manifest versions for release preparation."""

from __future__ import annotations

import argparse
import json
import re
from pathlib import Path


SEMVER_RE = re.compile(
    r"^(0|[1-9]\d*)\."
    r"(0|[1-9]\d*)\."
    r"(0|[1-9]\d*)"
    r"(?:-(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*)(?:\."
    r"(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*))*)?"
    r"(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$"
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Bump coding-plugins manifest versions.")
    parser.add_argument("version", help="Strict semver version, for example 0.6.16.")
    parser.add_argument("--root", default=".", help="Repository root. Defaults to the current directory.")
    return parser.parse_args()


def validate_version(version: str) -> None:
    if SEMVER_RE.fullmatch(version) is None:
        raise ValueError(f"Version must be strict semver: {version}")


def read_json(path: Path) -> dict[str, object]:
    return json.loads(path.read_text(encoding="utf-8"))


def write_json(path: Path, data: dict[str, object]) -> None:
    path.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def update_versions(root: Path, version: str) -> None:
    validate_version(version)
    root = root.resolve()
    config_path = root / ".version-bump.json"
    config = read_json(config_path)
    files = config.get("files")
    if not isinstance(files, list):
        raise ValueError(".version-bump.json must define a files array.")

    for entry in files:
        if not isinstance(entry, dict):
            raise ValueError(".version-bump.json files entries must be objects.")
        relative_path = entry.get("path")
        field = entry.get("field", "version")
        if not isinstance(relative_path, str) or not isinstance(field, str):
            raise ValueError(".version-bump.json files entries require string path and field.")
        path = root / relative_path
        data = read_json(path)
        data[field] = version
        write_json(path, data)


def main() -> int:
    args = parse_args()
    try:
        update_versions(Path(args.root), args.version)
    except (OSError, ValueError, json.JSONDecodeError) as error:
        print(f"Version bump failed: {error}")
        return 1

    print(f"Version bumped to {args.version}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/cli/release/bump-version.ts
#!/usr/bin/env node
import { readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";

const SEMVER_RE =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*)(?:\.(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*))*)?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/;

type JsonObject = Record<string, unknown>;

type VersionBumpFileEntry = {
  path?: unknown;
  field?: unknown;
};

function parseArgs(argv: string[]): { version: string; root: string } {
  const args = [...argv];
  const version = args.shift();
  if (!version) {
    throw new Error("the following arguments are required: version");
  }
  let root = ".";
  while (args.length > 0) {
    const option = args.shift();
    if (option === "--root") {
      const value = args.shift();
      if (!value) {
        throw new Error("argument --root: expected one argument");
      }
      root = value;
      continue;
    }
    throw new Error(`unrecognized arguments: ${option}`);
  }
  return { version, root };
}

function validateVersion(version: string): void {
  if (!SEMVER_RE.test(version)) {
    throw new Error(`Version must be strict semver: ${version}`);
  }
}

function readJson(path: string): JsonObject {
  return JSON.parse(readFileSync(path, "utf8")) as JsonObject;
}

function writeJson(path: string, data: JsonObject): void {
  writeFileSync(path, `${JSON.stringify(data, null, 2)}\n`, "utf8");
}

function updateVersions(root: string, version: string): void {
  validateVersion(version);
  const resolvedRoot = resolve(root);
  const configPath = resolve(resolvedRoot, ".version-bump.json");
  const config = readJson(configPath);
  const files = config.files;
  if (!Array.isArray(files)) {
    throw new Error(".version-bump.json must define a files array.");
  }

  for (const item of files as VersionBumpFileEntry[]) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      throw new Error(".version-bump.json files entries must be objects.");
    }
    const relativePath = item.path;
    const field = item.field ?? "version";
    if (typeof relativePath !== "string" || typeof field !== "string") {
      throw new Error(".version-bump.json files entries require string path and field.");
    }
    const targetPath = resolve(resolvedRoot, relativePath);
    const data = readJson(targetPath);
    data[field] = version;
    writeJson(targetPath, data);
  }
}

function main(argv: string[]): number {
  try {
    const args = parseArgs(argv);
    updateVersions(args.root, args.version);
    console.log(`Version bumped to ${args.version}`);
    return 0;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.log(`Version bump failed: ${message}`);
    return 1;
  }
}

process.exitCode = main(process.argv.slice(2));
